File: Topics/01_Encapsulation/01_02_Classes_and_objects/cpp_source/using.cpp
#include <iostream>
#include <string>
using namespace std;

class Clock {

    int hours, minutes, seconds;

public:

    Clock(int h, int m, int s) :
            hours(h), minutes(m), seconds(s) {
    }

    virtual int getHours() {
        return hours;
    }

};

class Animal {
public:
    virtual void sayHi() {
        cout << "I am not a plant." << endl;
    }
};

class Dog: public Animal {
public:
    virtual void sayHi() {
        cout << "Bark Bark!" << endl;
    }
};

class RESTClient {
public:
    RESTClient(string url) {
    }

    string get(string url) {
        return "";

    }

    void post(string url, string data) {
    }
};

class Robot {

    int location;
    double battery;

    void moveLeftLeg() {
    }
    void moveRightLeg() {
    }

public:

    int getLocation() {
        return location;
    }

    void goTo(int loc) {
        location = loc;
        battery = battery - 0.10;
    }

    void goTo2(Robot * that, int location) {
        this->location = location;
    }

    void sayHi() {
        cout << "Beep Beep" << endl;
    }
    double getBatteryLevel() {
        return 0.0;
    }
    void goTo(double lat, double lng) {
        moveLeftLeg();
        moveRightLeg();
    }
};

class Point {
    int x, y;
public:
    int addX(int a) {
        return x + a;
        return this->x + a;
    }
};

int main() {

    Point * p = new Point();
    int a = p->addX(5);

    Point q;
    int b = q.addX(5);

    Robot * bob = new Robot();
    bob->goTo(20);
    int a = bob->getLocation();

    Robot * jan = new Robot();
    jan->goTo(50);
    int b = bob->getLocation();

    Robot * rob = new Robot();
    rob->sayHi();
    double charge = rob->getBatteryLevel();
    cout << charge << endl;
    rob->goTo(34.738, -86.60);

    //rob->moveLeftLeg();

    /*
     RESTClient * client = new RESTClient("https://io.adafruit.com/topher");

     string resp = client->get("/freezer_data");

     client->post("/freezer_data", "-10.5");

     delete client;

     RESTClient client2("https://io.adafruit.com/topher");
     string resp = client2.get("/freezer_data");
     client2.post("/freezer_data", "-10.5");

     cout << "HERE" << endl;
     */
}

File: Topics/01_Encapsulation/01_02_Classes_and_objects/python_source/using.py


class Robot:
    
    def goTo(self,location):
        self.location = location
        battery = battery - 0.10;
        
    def get_location(self):
        return self._location
    
    def _move_left_leg(self):
        pass
    
    def _move_right_leg(self):
        pass
    
    def say_hi(self):
        print('Beep Beep')
        
    def get_battery_level(self):
        return 0.0
    
    def go_to(self,lat: float,lng: float=0) -> None:
        self._move_left_leg()
        self._move_right_leg()
        
class Point:
    
    def addX(self,a):
        return self.x + a
    
p = Point()
a = p.add(5)
        
bob = Robot()
bob.go_to(20)
x = bob.get_location()

jan = Robot()
jan._go_to(50)
y = jan.get_location()

bob._location = 100
        
rob = Robot()

rob.say_hi()

lev = rob.get_battery_level()

rob.go_to(34.738, -86.60)

rob._move_left_leg()


import logging

logger = logging.getLogger()

logger.setLevel(logging.DEBUG)

logger.warning('Bad things!')
logger.debug('Starting up')

b = logger.isEnabledFor(logging.WARN)
print(b)